pip/index.py

```diff
@@ -779,8 +791,8 @@ def _link_package_versions(self, link, search_name):
                     link, 'Python version is incorrect')
                 return
         logger.debug('Found link %s, version: %s', link, version)
-
-        return InstallationCandidate(search_name, version, link)
+        link.canonical_name = search.canonical
```

Comment:
Not a big fan of this extra attribute on `Link`.
A new argument could be passed instead to `cached_wheel` since it is called from `InstallRequirement.link.setter`